gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,216 @@
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching.v2;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Queues up the elements until {@link #flush()} is called; once batching is over, returned future
+ * resolves.
+ *
+ * <p>This class is not thread-safe, and expects to be used from a single thread.
+ */
+@BetaApi("The surface for batching is not stable yet and may change in the future.")
+@InternalApi
+public class BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT>
+    implements Batcher<ElementT, ElementResultT> {
+
+  private final BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT>
+      batchingDescriptor;
+  private final UnaryCallable<RequestT, ResponseT> callable;
+  private final RequestT prototype;
+  private Batch<ElementT, ElementResultT, RequestT> currentOpenBatch;
+
+  private final AtomicInteger numOfOutstandingBatches = new AtomicInteger(0);
+  private final Object flushLock = new Object();
+  private boolean isClosed = false;
+
+  private BatcherImpl(Builder<ElementT, ElementResultT, RequestT, ResponseT> builder) {
+    this.prototype = checkNotNull(builder.prototype, "RequestPrototype cannot be null.");
```

Comment:
Please do not put period (`.`) at the end of messages like this to stay consistent with the rest of the codebase (applies to other places as well, like line 113).

Also the message probably should not use the class names to indicate what is null (either use field names i.e `prototype` instead of `RequestPrototype`, or regular langugae to describe what is null, otherwise it looks like the "class" itself is null).